src/effects/RealtimeEffectStateUI.cpp

```diff
@@ -0,0 +1,266 @@
+};
+} // namespace
+
+
+// A shared state, that manages the UI
+class RealtimeEffectStateUI::SharedState final
+{
+public:
+   ~SharedState()
+   {
+      Hide();
+   }
+
+   void SetActiveState(RealtimeEffectState& state)
+   {
+      mCurrentState = state.shared_from_this();
+
+      auto prevInstance = mCurrentInstance.lock();
+
+      if (prevInstance != nullptr && IsShown())
+      {
+         auto instance = state.GetInstance();
+
+         // In case of instance mismatch we have to close the dialog
+         if (instance != prevInstance)
+         {
+            // However, the instance is kept by the dialog
+            // as a strong pointer and is copied between RealTimeState
+            // as a weak one. Instance mismatch means that something
+            // really unexpected has happened.
+            assert(instance == prevInstance);
+            // In release builds - the dialog will be jsut closed
+            Hide();
+         }
+      }
+   }
+
+   bool IsShown() const noexcept
+   {
+      return mEffectUIHost != nullptr;
+   }
+
+   void Show(AudacityProject& project)
+   {
+      if (mEffectUIHost != nullptr && mEffectUIHost->IsShown())
+      {
+         // Bring the host to front
+         mEffectUIHost->Raise();
+         return;
+      }
+
+      auto currentState = mCurrentState.lock();
+
+      // Active state is set in RealtimeEffectStateUI::Get
+      // It is an unexpected situation when it has expired before
+      // Show is called.
+      assert(currentState != nullptr);
+
+      if (currentState == nullptr)
+         return;
+
+      const auto ID = currentState->GetID();
+      const auto effectPlugin = EffectManager::Get().GetEffect(ID);
+
+      EffectUIClientInterface* client =
+         effectPlugin->GetEffectUIClientInterface();
+
+      // Effect has no client interface
+      if (client == nullptr)
+         return;
+
+      // mUIHost will acquire the correct instance
+      // and will manage it's lifetime
+      wxWindow& projectWindow = ProjectWindow::Get(project);
+      
+      std::shared_ptr<EffectInstance> pInstance;
+
+      auto access = currentState->GetAccess();
+
+      // EffectUIHost invokes shared_from_this on access
+      Destroy_ptr<EffectUIHost> dlg(safenew EffectUIHost(
+         &projectWindow, project, *effectPlugin, *client, pInstance, *access,
+         currentState));
+
+      if (!dlg->Initialize())
+         return;
+
+      // pInstance is updated in EffectUIHost constructor  
+      mCurrentInstance = pInstance;
+      // Dialog is owned by it's parent now!
+      mEffectUIHost = dlg.release();
+
+      UpdateTitle();
+
+      client->ShowClientInterface(
+         projectWindow, *mEffectUIHost, mEffectUIHost->GetValidator(), false);
+
+      // The dialog was modal? That shouldn't have happened
+      if (mEffectUIHost == nullptr || !mEffectUIHost->IsShown())
+      {
+         assert(mEffectUIHost != nullptr && mEffectUIHost->IsShown());
```

Comment:
or just `assert(false)`

Reply:
Ironically, there was assert(false), but at some point I have decided to make it more explicit